File: linked-list/21.Merge-Two-Sorted-Lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # Time complexity is O(m + n) since each node from both lists is processed exactly once.
        # Space complexity is O(m + n) because the recursive call stack can grow to the total number of nodes in the worst case.

        # We have to note that this way modifies the original lists rather than creating new nodes, if we wanted to not mutate the arguments then we'll just create dummy values.

        if list1 is None:
            return list2
        elif list2 is None: 
            return list1

        if list1.val > list2.val:
            list2.next = self.mergeTwoLists(list1, list2.next)
            return list2
        
        else:
            list1.next = self.mergeTwoLists(list1.next, list2)
            return list1

File: linked-list/21.Merge-Two-Sorted-Lists/merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        // time complexity is O(m + n) as shown in python file
        // we want to avoid doing this recursively because the recursive stack will get big.
        // how can we do this in O(1) space complexity? use dummy node as the first and return
        // dummy.next
        if (list1 == null) {
            return list2;
        }

        else if (list2 == null) {
            return list1;
        }

        ListNode dummy = new ListNode();
        ListNode head = dummy;

        while (list1 != null && list2 != null) {
            if (list1.val > list2.val) {
                dummy.next = list2;
                list2 = list2.next;
            } else {
                dummy.next = list1;
                list1 = list1.next;
            }
            dummy = dummy.next;
        }

        if (list1 != null) {
            dummy.next = list1;
        } else {
            dummy.next = list2;
        }

        return head.next;

        
    }
}
